lua/core/nvimtree.lua

```diff
@@ -73,63 +73,36 @@ M.setup = function()
       { key = { "l", "<CR>", "o" }, cb = tree_cb "edit" },
       { key = "h", cb = tree_cb "close_node" },
       { key = "v", cb = tree_cb "vsplit" },
-      { key = "q", cb = ":lua require('core.nvimtree').toggle_tree()<cr>" },
     }
   end
-end
---
-M.focus_or_close = function()
-  local view_status_ok, view = pcall(require, "nvim-tree.view")
-  if not view_status_ok then
-    return
-  end
-  local a = vim.api
 
-  local curwin = a.nvim_get_current_win()
-  local curbuf = a.nvim_win_get_buf(curwin)
-  local bufnr = view.View.bufnr
-  local winnr = view.get_winnr()
+  lvim.builtin.which_key.mappings["e"] = { "<cmd>NvimTreeToggle<CR>", "Explorer" }
 
-  if view.win_open() then
-    if curwin == winnr and curbuf == bufnr then
-      view.close()
-      if package.loaded["bufferline.state"] then
-        require("bufferline.state").set_offset(0)
-      end
-    else
-      view.focus()
-    end
-  else
-    view.open()
-    if package.loaded["bufferline.state"] and lvim.builtin.nvimtree.side == "left" then
-      -- require'bufferline.state'.set_offset(lvim.builtin.nvimtree.width + 1, 'File Explorer')
-      require("bufferline.state").set_offset(lvim.builtin.nvimtree.width + 1, "")
-    end
-  end
+  vim.cmd "au FileType NvimTree lua require('core.nvimtree').on_type()"
 end
 --
-M.toggle_tree = function()
-  local view_status_ok, view = pcall(require, "nvim-tree.view")
-  if not view_status_ok then
-    return
+M.on_type = function()
+  local buf = vim.fn.expand "<abuf>"
+  vim.cmd("au BufWinEnter <buffer=" .. buf .. "> lua require('core.nvimtree').on_open()")
+  vim.cmd("au WinClosed   <buffer=" .. buf .. "> lua require('core.nvimtree').on_close()")
+end
```

Comment:
Is the use-case worth adding 3 additional auto-commands trying to fix a limitation in nvim-tree? are there no performance penalties for using this [vimL] ? I don't think anyone uses `NvimTreeOpen`. And I'm not sure what the problem is with `:q`.

I feel like all of this can be just as easily added in `config.lua`.

Reply:
And honestly for peace of mind, I don't mind that

Reply:
By the way, I still don't fully get the problem with `:q`.

So nvim-tree doesn't handle the bufferline, that's up to barbar which lists the following as a way to handle this: https://github.com/romgrk/barbar.nvim#integration-with-filetree-plugins

> And honestly for peace of mind, I don't mind that

How is it a peace of mind to add a hook that gets evaluated every time you open nvim-tree, but the same hook is not actually maintained by nvim-tree?

Reply:
The problem is not `:q` nor is it `:NvimTreeOpen`, the problem is a combination of everything that can open or close nvim-tree without calling `M.toggle_tree`.

Reply:
> The problem is not `:q` nor is it `:NvimTreeOpen`, the problem is a combination of everything that can open or close nvim-tree without calling `M.toggle_tree`.

I still don't know what the _problem that we're trying to fix_ is. My thought is that you don't have to use that binding, and just use the default ones instead.  

```lua
-- just add this to you config.lua
lvim.builtin.which_key.mappings["e"] = { "<cmd>NvimTreeToggle<CR>", "Explorer" }
```

> But if you want this to be inside the nvim-tree codebase, then sure, I could create an issue there for them to create a callback.

If it's an **nvim-tree** problem that's not related to **barbar** specifically, then yeah I would say it's a better fit there instead.

Reply:
> I still don't know what the _problem that we're trying to fix_ is. 

I was pretty clear, but let me show you instead.

https://user-images.githubusercontent.com/36672196/130056420-69a5a43d-b238-4778-9cd4-abc8547bc48e.mp4

Reply:
That's related to Dashboard and not nvim-tree. Try to use the "glepnir/dashboard-nvim" instead of Chris' fork and you'll see that the problem is no longer there. I'm not sure if the fork is still necessary btw.

![Xzz4gsFdAh](https://user-images.githubusercontent.com/59826753/130087263-9740f59f-9add-4ce9-9547-7ed2daeaf9df.gif)

What about the `:q` issue?

Reply:
Dude... wtf. I don't care if the tabs show or hide on dhasboard. My point is the same offset bug is visible on a normal file...